pppd/sys-linux.c

```diff
@@ -1491,6 +1494,118 @@ get_ppp_stats_ioctl(int u, struct pppd_stats *stats)
     return 1;
 }
 
+/********************************************************************
+ * get_ppp_stats_rtnetlink - return statistics for the link, using rtnetlink
+ * This provides native 64-bit counters.
+ */
+#ifdef RTM_GETSTATS
+static int
+get_ppp_stats_rtnetlink(int u, struct pppd_stats *stats)
+{
+    static int rtnl_fd = -1;
+
+    struct sockaddr_nl nladdr;
+    struct {
+        struct nlmsghdr nlh;
+        struct if_stats_msg ifsm;
+    } nlreq;
+    struct {
+        struct nlmsghdr nlh;
+	union {
+	    struct nlmsgerr nlerr;
+	    struct {
+		struct rtmsg rth;
+		struct rtnl_link_stats64 stats;
+	    };
+	};
+    } nlresp;
+    ssize_t nlresplen;
+    struct iovec iov;
+    struct msghdr msg;
+
+    if (rtnl_fd < 0) {
+	rtnl_fd = socket(AF_NETLINK, SOCK_RAW, NETLINK_ROUTE);
+	if (rtnl_fd < 0) {
+	    error("get_ppp_stats_rtnetlink: error creating NETLINK socket: %m (line %d)", __LINE__);
+	    return 0;
+	}
+
+	memset(&nladdr, 0, sizeof(nladdr));
+	nladdr.nl_family = AF_NETLINK;
+	nladdr.nl_pid = getpid();
+
+	if (bind(rtnl_fd, (struct sockaddr *)&nladdr, sizeof(nladdr)) < 0) {
+	    error("get_ppp_stats_rtnetlink: bind(AF_NETLINK): %m (line %d)", __LINE__);
+	    goto err;
+	}
+    }
+
+    memset(&nladdr, 0, sizeof(nladdr));
+    nladdr.nl_family = AF_NETLINK;
+
+    memset(&nlreq, 0, sizeof(nlreq));
+    nlreq.nlh.nlmsg_len = sizeof(nlreq);
+    nlreq.nlh.nlmsg_type = RTM_GETSTATS;
+    nlreq.nlh.nlmsg_flags = NLM_F_REQUEST;
+
+    nlreq.ifsm.ifindex = if_nametoindex(ifname);
+    nlreq.ifsm.filter_mask = IFLA_STATS_LINK_64;
+
+    memset(&iov, 0, sizeof(iov));
+    iov.iov_base = &nlreq;
+    iov.iov_len = sizeof(nlreq);
+
+    memset(&msg, 0, sizeof(msg));
+    msg.msg_name = &nladdr;
+    msg.msg_namelen = sizeof(nladdr);
+    msg.msg_iov = &iov;
+    msg.msg_iovlen = 1;
+
+    if (sendmsg(rtnl_fd, &msg, 0) < 0) {
+        error("get_ppp_stats_rtnetlink: sendmsg(RTM_GETSTATS): %m (line %d)", __LINE__);
+	goto err;
+    }
+
+    /* We just need to repoint to IOV ... everything else stays the same */
+    iov.iov_base = &nlresp;
+    iov.iov_len = sizeof(nlresp);
+
+    nlresplen = recvmsg(rtnl_fd, &msg, 0);
+
+    if (nlresplen < 0) {
+        error("get_ppp_stats_rtnetlink: recvmsg(RTM_GETSTATS): %m (line %d)", __LINE__);
+	goto err;
+    }
+
+    if (!NLMSG_OK(&nlresp.nlh, nlresplen)) {
+	error("get_ppp_stats_rtnetlink: Netlink response message is corrupt/incomplete (line %d)", __LINE__);
+	goto err;
+    }
+
+    if (nlresp.nlh.nlmsg_type == NLMSG_ERROR) {
+	error("get_ppp_stats_rtnetlink: Netlink responded with error: %s (line %d)", strerror(-nlresp.nlerr.error), __LINE__);
+	goto err;
+    }
+
+    if (nlresp.nlh.nlmsg_type != RTM_NEWSTATS) {
+	error("get_ppp_stats_rtnetlink: Expected RTM_NEWSTATS response, found something else (mlmsg_type %d, line %d)",
+		nlresp.nlh.nlmsg_type, __LINE__);
+	goto err;
+    }
+
```

Comment:
NLMSG_OK ... right, looking at the headers what you say makes sense.

So before accessing the sub fields need to check that adequate data has come with the packet (which I'm fairly certain we can assume).  Since the fields we care about are the first four in the stats structure, and the structure in theory can grow over time ... do I just check that the message covers those first four fields, or do I check that we receive the full struct?

```sizeof(struct nlmsg)=16
sizeof(struct rtmsg)=12
sizeof(struct rtnl_link_stats64)=192
sizeof(nlresp)=224
RTM_PAYLOAD(nlresp.nlh) size=196
msg len=224```

Looking at the header for rtnetlink the RTA_OK seems to be the closest thing to NLMSG_OK, and this is for additional attributes not sure it'll fly for this message (There is no rtattr struct).

So I believe the following checks should be adequate:

1.  Before accessing nlmsgerr check that NLMSG_PAYLOAD(&nlh, 0) >= sizeof(nlerr).
2.  Before accessing stats check that RTM_PAYLOAD(&nlh) >= sizeof(stats).

Two works because RTM_PAYLOAD is defined in terms of NLMSG_PAYLOAD.

Reply:
Check for first four fields should be enough. Also it should be possible to instruct kernel to just send to userspace only truncated buffer with these first four fields.

Reply:
Done and as side effect the messages from kernel space is not truncated too ...